src/profile-logic/profile-data.js

```diff
@@ -3371,6 +3375,136 @@ export function findAddressProofForFile(
+ * cases, a call node can have its native code in multiple different functions,
+ * for example if it was inlined into multiple different functions.
+ * Another way to get multiple native symbols here is if a call tree transform
+ * collapsed frames from different original functions into a merged function.
+ */
+export function getNativeSymbolsForCallNode(
+  callNodeIndex: IndexIntoCallNodeTable,
+  { stackIndexToCallNodeIndex }: CallNodeInfo,
+  stackTable: StackTable,
+  frameTable: FrameTable
+): Set<IndexIntoNativeSymbolTable> {
+  const set = new Set();
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    if (stackIndexToCallNodeIndex[stackIndex] === callNodeIndex) {
+      const frame = stackTable.frame[stackIndex];
+      const nativeSymbol = frameTable.nativeSymbol[frame];
+      if (nativeSymbol !== null) {
+        set.add(nativeSymbol);
+      }
+    }
+  }
+  return set;
+}
+
+/**
+ * Convert a native symbol index into a NativeSymbolInfo object, to create
+ * something that's meaningful outside of its associated thread.
+ */
+export function getNativeSymbolInfo(
+  nativeSymbol: IndexIntoNativeSymbolTable,
+  nativeSymbols: NativeSymbolTable,
+  frameTable: FrameTable,
+  stringTable: UniqueStringArray
+): NativeSymbolInfo {
+  const functionSizeOrNull = nativeSymbols.functionSize[nativeSymbol];
+  const functionSize =
+    functionSizeOrNull ??
+    calculateFunctionSizeLowerBound(
+      frameTable,
+      nativeSymbols.address[nativeSymbol],
+      nativeSymbol
+    );
+  return {
+    libIndex: nativeSymbols.libIndex[nativeSymbol],
+    address: nativeSymbols.address[nativeSymbol],
+    name: stringTable.getString(nativeSymbols.name[nativeSymbol]),
+    functionSize,
+    functionSizeIsKnown: functionSizeOrNull !== null,
+  };
+}
+
+/**
+ * Calculate the BottomBoxInfo for a call node, i.e. information about which
+ * things should be shown in the profiler UI's "bottom box" when this call node
+ * is double-clicked.
+ *
+ * We always want to update all panes in the bottom box when a new call node is
+ * double-clicked, so that we don't show inconsistent information side-by-side.
+ */
+export function getBottomBoxInfoForCallNode(
+  callNodeIndex: IndexIntoCallNodeTable,
+  callNodeInfo: CallNodeInfo,
+  thread: Thread
+): BottomBoxInfo {
+  const {
+    stackTable,
+    frameTable,
+    funcTable,
+    stringTable,
+    resourceTable,
+    nativeSymbols,
+  } = thread;
+
+  const funcIndex = callNodeInfo.callNodeTable.func[callNodeIndex];
+  const fileName = funcTable.fileName[funcIndex];
+  const sourceFile = fileName !== null ? stringTable.getString(fileName) : null;
+  const resource = funcTable.resource[funcIndex];
+  const libIndex =
+    resource !== -1 && resourceTable.type[resource] === resourceTypes.library
+      ? resourceTable.lib[resource]
+      : null;
+
+  const nativeSymbolSet = new Set(
+    [
+      ...getNativeSymbolsForCallNode(
```

Comment:
Can the result of `getNativeSymbolsForCallNode` have a lot of items? My understanding is that this is very unlikely, but could that happen for inverted trees for example? In that case would it make sense to iterate the Set with (for example) a `for .. of` and create the new one with `set.add`?

On another hand, it doesn't look like that having a Set for `nativeSymbolSet` is useful since it contains objects. What do you think of using an Array for this data?

Reply:
In the vast majority of cases it will be 1 element, sometimes 2 or 3, and probably never more than 100 (though you could create a synthetic testcase with an arbitrary amount).

Using an array sounds good to me!